WIP: sql: add int[] column type

cuongdo opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10584)
<!-- Reviewable:end -->

Commit:
sql: support int[] Type, Datum, and column type

The main parts are:
1. Extend `tArray` to properly support different types.
2. Change the `DArray` implementation to store the corresponding `Type`,
   to correctly implement the `Datum` interface.
3. Support the int[] column type.  This is currently meant to be used
   only for virtual schemas such as `pg_catalog`.

cuongdo commented:
This a WIP and has still-broken tests. I mainly want know if I'm on the right track with regard to the implementation of the `int[]` column type. After I get some feedback, I'll fix the tests.

nvanbenschoten commented:
This definitely seems like it's moving in the right direction. One of the biggest things to focus on here is that we're going to have to be smart about how we introduce arrays into the `sqlbase.ColumnType` protobuf. I'd assume that means relying on recursion for both the definition and some of the sourounding functions.

---

Reviewed 7 of 8 files at r1, 3 of 8 files at r2, 1 of 1 files at r3.
Review status: 11 of 16 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[pkg/sql/pg_catalog.go, line 381 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KWBUsShCBDE2JfwlZ3Y:-KWBUsShCBDE2JfwlZ3Z:bjg5hk0) ([raw file](https://github.com/cockroachdb/cockroach/blob/712cc60a0f7506bca289d9a61f8efb5ee8c183bd/pkg/sql/pg_catalog.go#L381)):*
> ```Go
> 	coninhcount INT,
> 	connoinherit BOOL,
> 	conkey INT[],
> ```

That's a pretty sight :+1: 

---

*[pkg/sql/pg_catalog.go, line 504 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KWBUoEBUyL5o1oxcnSg:-KWBUoEBUyL5o1oxcnSh:b-kjcvc3) ([raw file](https://github.com/cockroachdb/cockroach/blob/712cc60a0f7506bca289d9a61f8efb5ee8c183bd/pkg/sql/pg_catalog.go#L504)):*
> ```Go
> 		return parser.DNull
> 	}
> 	d := parser.DArray{Typ: parser.TypeIntArray}
> ```

I'd use the new constructor here

---

*[pkg/sql/pg_catalog.go, line 1138 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KWBUj1c83_iHU92vfzP:-KWBUj1c83_iHU92vfzQ:b-un059z) ([raw file](https://github.com/cockroachdb/cockroach/blob/712cc60a0f7506bca289d9a61f8efb5ee8c183bd/pkg/sql/pg_catalog.go#L1138)):*
> ```Go
> var datumToTypeCategory = map[reflect.Type]*parser.DString{
> 	reflect.TypeOf(parser.TypeAny):         typCategoryPseudo,
> 	reflect.TypeOf(parser.TypeStringArray): typCategoryArray,
> ```

Does this actually work?

---

*[pkg/sql/parser/builtins.go, line 969 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KWBUMSnAVZzXWqCYFVW:-KWBUMSnAVZzXWqCYFVX:bvjkj6w) ([raw file](https://github.com/cockroachdb/cockroach/blob/712cc60a0f7506bca289d9a61f8efb5ee8c183bd/pkg/sql/parser/builtins.go#L969)):*
> ```Go
> 				if showImplicitSchemas == DBoolTrue {
> 					for _, p := range ctx.SearchPath {
> 						schemas.Array = append(schemas.Array, NewDString(p))
> ```

This is such a common operation that I wouldn't object to putting it in a method, especially if we included a parameterized type sanity check.

---

*[pkg/sql/parser/col_types.go, line 256 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KWBVM4PGANmEm1VYKqi:-KWBVM4PGANmEm1VYKqj:bh7l0sr) ([raw file](https://github.com/cockroachdb/cockroach/blob/712cc60a0f7506bca289d9a61f8efb5ee8c183bd/pkg/sql/parser/col_types.go#L256)):*
> ```Go
> 	Name string
> 	// Typ is the type of the elements in this array.
> 	Typ ColumnType
> ```

same comment about parametrized type wording.

---

*[pkg/sql/parser/col_types.go, line 268 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KWBVPb4Ah2_z7c8ObW2:-KWBVPb5klme7LY0gImx:bb7lbyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/712cc60a0f7506bca289d9a61f8efb5ee8c183bd/pkg/sql/parser/col_types.go#L268)):*
> ```Go
> 		return arrayColTypeIntArray
> 	}
> 	return nil
> ```

nil?

---

*[pkg/sql/parser/datum.go, line 1346 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KWBTxtVzwmSf3Ld41za:-KWBTxtVzwmSf3Ld41zb:bfft3z3) ([raw file](https://github.com/cockroachdb/cockroach/blob/712cc60a0f7506bca289d9a61f8efb5ee8c183bd/pkg/sql/parser/datum.go#L1346)):*
> ```Go
> // text during serialization.
> type DArray struct {
> 	Typ   Type
> ```

Throughout: I'd favor the term "parameterized type" (`paramTyp` for short?) over "type"

---

*[pkg/sql/parser/sql.y, line 400 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KWBVxN1NhEpqax2RXq8:-KWBVxN1NhEpqax2RXq9:b-4cgo0j) ([raw file](https://github.com/cockroachdb/cockroach/blob/712cc60a0f7506bca289d9a61f8efb5ee8c183bd/pkg/sql/parser/sql.y#L400)):*
> ```Yacc
> %type <IndexElemList> index_params
> %type <NameList> name_list opt_name_list
> %type <bool> opt_array_bounds
> ```

It looks like this is going to be an `Exprs` type when we fix the unimplemented array with length rule below (see https://github.com/postgres/postgres/blob/8c48375e5f43ebd832f93c9166d1fe0e639ff806/src/backend/parser/gram.y). Let's implement it that way right off the bat. For now, we can just check if the `Exprs` are nil or not in place of `$2.bool()`.

---

*[pkg/sql/parser/type.go, line 255 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KWBTTEY4mXLcZVjSTm7:-KWBTTEY4mXLcZVjSTm8:b-glbpvw) ([raw file](https://github.com/cockroachdb/cockroach/blob/712cc60a0f7506bca289d9a61f8efb5ee8c183bd/pkg/sql/parser/type.go#L255)):*
> ```Go
> 		return false
> 	}
> 	return a.Typ == oa.Typ
> ```

`a.Typ.Equal(oa.Typ)`! This needs to work recursively

---

*[pkg/sql/sqlbase/structured.proto, line 41 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KWBXBUZulFHdV1s0y-5:-KWBXBUZulFHdV1s0y-6:byt8yxx) ([raw file](https://github.com/cockroachdb/cockroach/blob/712cc60a0f7506bca289d9a61f8efb5ee8c183bd/pkg/sql/sqlbase/structured.proto#L41)):*
> ```protobuf
>     BYTES = 8;
>     TIMESTAMPTZ = 9;
>     INT_ARRAY = 10;
> ```

I think we need to implement this in a way that can be recursively applied. Imagine supporting the column type `INT[][][]`. I'd picture ColumnType somehow having an "array" option that holds a ColumnType itself.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-:-KWBY5oDzQXUuWHwNS-b:b-w6e6k9)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
*[pkg/sql/parser/sql.y, line 400 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KWBVxN1NhEpqax2RXq8:-KWYhkzrQdcjETakQq58:b-ok8l1m) ([raw file](https://github.com/cockroachdb/cockroach/blob/712cc60a0f7506bca289d9a61f8efb5ee8c183bd/pkg/sql/parser/sql.y#L400)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> It looks like this is going to be an `Exprs` type when we fix the unimplemented array with length rule below (see https://github.com/postgres/postgres/blob/8c48375e5f43ebd832f93c9166d1fe0e639ff806/src/backend/parser/gram.y). Let's implement it that way right off the bat. For now, we can just check if the `Exprs` are nil or not in place of `$2.bool()`.
</details>
I don't understand your comment (though that's probably due to my rustiness with YACC). I'm looking at these:

https://github.com/postgres/postgres/blob/8c48375e5f43ebd832f93c9166d1fe0e639ff806/src/backend/parser/gram.y#L364
https://github.com/postgres/postgres/blob/8c48375e5f43ebd832f93c9166d1fe0e639ff806/src/backend/parser/gram.y#L11158

and seeing that in the Postgres grammar, `opt_array_bounds` are simply lists of integers. Or, am I misinterpreting something here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
*[pkg/sql/sqlbase/structured.proto, line 41 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KWBXBUZulFHdV1s0y-5:-KWZKXfg9beN6N8pW4QZ:b-cogmrl) ([raw file](https://github.com/cockroachdb/cockroach/blob/712cc60a0f7506bca289d9a61f8efb5ee8c183bd/pkg/sql/sqlbase/structured.proto#L41)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> I think we need to implement this in a way that can be recursively applied. Imagine supporting the column type `INT[][][]`. I'd picture ColumnType somehow having an "array" option that holds a ColumnType itself.
</details>
Why does this need to be recursive? It seems like Postgres requires that sub-arrays have matching dimensions. For example:

```
postgres=# INSERT INTO sal_emp VALUES ('foo', '{3,5}', '{{"foo"},{"foo", "bar"}}');
ERROR:  malformed array literal: "{{"foo"},{"foo", "bar"}}"
LINE 1: INSERT INTO sal_emp VALUES ('foo', '{3,5}', '{{"foo"},{"foo"...
                                                    ^
DETAIL:  Multidimensional arrays must have sub-arrays with matching dimensions.
```

So, instead of what I currently have, wouldn't the following new proto fields be sufficient for fully specifying an array?

```
optional Kind arrayKind = 4;
repeated uint32 arrayDimensions = 5;
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 11 of 16 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[pkg/sql/sqlbase/structured.proto, line 41 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KWBXBUZulFHdV1s0y-5:-KWZqVzFM0JzLXKdXRn9:bkrqifp) ([raw file](https://github.com/cockroachdb/cockroach/blob/712cc60a0f7506bca289d9a61f8efb5ee8c183bd/pkg/sql/sqlbase/structured.proto#L41)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> Why does this need to be recursive? It seems like Postgres requires that sub-arrays have matching dimensions. For example:
> 
> ```
> postgres=# INSERT INTO sal_emp VALUES ('foo', '{3,5}', '{{"foo"},{"foo", "bar"}}');
> ERROR:  malformed array literal: "{{"foo"},{"foo", "bar"}}"
> LINE 1: INSERT INTO sal_emp VALUES ('foo', '{3,5}', '{{"foo"},{"foo"...
>                                                     ^
> DETAIL:  Multidimensional arrays must have sub-arrays with matching dimensions.
> ```
> 
> So, instead of what I currently have, wouldn't the following new proto fields be sufficient for fully specifying an array?
> 
> ```
> optional Kind arrayKind = 4;
> repeated uint32 arrayDimensions = 5;
> ```
</details>
Ok I want to react to this point specifically.
There is no doubt that what we want is (like pg) homogeneous arrays, where all elements have the same type and "every row has the same number of columns" (the general property is to be multi-dimensional rectangularity, i.e. each dimension has the same size at every position in the other dimensions).

To achieve this, there are two implementation directions. 
In both cases, there is a single array in memory that carries all the values.
The difference is how you store the "metadata".

One direction is to have a single descriptor containing two pointers; one pointer to the full value array and one pointer to an array of sizes, one per dimension. 

The other direction is to have nested descriptors; each descriptor corresponds to a "type"; each descriptor contains 3 things, 1) a pointer to a slice, 2) a pointer to the type of the elements (which can be another array type), and a *stride*: an integer value used by the indexing function to multiply by a coordinate at that level to access one element in the slice.

The advantage of the first method is that it is slightly cheaper in memory usage when you have a large number of dimensions.
The advantage of the second method is that the code for indexing is simpler.

Time-complexity wise they come to about the same. I think at this point it is up to the person who will implement this to decide how to do it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
Just submitted another set of changes. The biggest remaining open item are the changes to `structured.proto`.

Some tests are probably still failing, and I'll get to those after resolving the proto stuff.

---

Review status: 4 of 16 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[pkg/sql/pg_catalog.go, line 504 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KWBUoEBUyL5o1oxcnSg:-KWZtIdcMP_Z8Orn3_ba:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/712cc60a0f7506bca289d9a61f8efb5ee8c183bd/pkg/sql/pg_catalog.go#L504)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> I'd use the new constructor here
</details>
Done.

---

*[pkg/sql/pg_catalog.go, line 1138 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KWBUj1c83_iHU92vfzP:-KWE5pfAcDPWDSrJJgYQ:b3279j2) ([raw file](https://github.com/cockroachdb/cockroach/blob/712cc60a0f7506bca289d9a61f8efb5ee8c183bd/pkg/sql/pg_catalog.go#L1138)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Does this actually work?
</details>
It seems to, though I'm not clear what the symptoms of this *not* working would be.

---

*[pkg/sql/parser/builtins.go, line 969 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KWBUMSnAVZzXWqCYFVW:-KWZgB2A0BKuZooKRBgt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/712cc60a0f7506bca289d9a61f8efb5ee8c183bd/pkg/sql/parser/builtins.go#L969)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> This is such a common operation that I wouldn't object to putting it in a method, especially if we included a parameterized type sanity check.
</details>
Done.

---

*[pkg/sql/parser/col_types.go, line 256 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KWBVM4PGANmEm1VYKqi:-KWZgQkCMMD2OISrKMmy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/712cc60a0f7506bca289d9a61f8efb5ee8c183bd/pkg/sql/parser/col_types.go#L256)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> same comment about parametrized type wording.
</details>
Done.

---

*[pkg/sql/parser/col_types.go, line 268 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KWBVPb4Ah2_z7c8ObW2:-KWYenPKuDbpdNd_I5QM:b1jcqvm) ([raw file](https://github.com/cockroachdb/cockroach/blob/712cc60a0f7506bca289d9a61f8efb5ee8c183bd/pkg/sql/parser/col_types.go#L268)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> nil?
</details>
I'm not sure what reasonable alternative there is to `nil`. `DatumTypeToColumnType`, in the same file, returns `nil` upon error. However, I've added an `error` return value to supplement the `nil`.

---

*[pkg/sql/parser/datum.go, line 1346 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KWBTxtVzwmSf3Ld41za:-KWZggIGTtxQ1rnvaDRH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/712cc60a0f7506bca289d9a61f8efb5ee8c183bd/pkg/sql/parser/datum.go#L1346)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Throughout: I'd favor the term "parameterized type" (`paramTyp` for short?) over "type"
</details>
Done.

---

*[pkg/sql/parser/type.go, line 255 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KWBTTEY4mXLcZVjSTm7:-KWZhjFwZ2z1iTP35WVc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/712cc60a0f7506bca289d9a61f8efb5ee8c183bd/pkg/sql/parser/type.go#L255)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> `a.Typ.Equal(oa.Typ)`! This needs to work recursively
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-:-KWZtJaxMydHNnVRTXGx:b-kdxsst)*
<!-- Sent from Reviewable.io -->

knz commented:
For the protobuf I'd recommend the "simple descriptor" approach (see my other comment) with two fields: one array of sizes (one per dimension) and one for the element type. 

Nathan: the "recursive type" problem is only relevant when we start supporting product types.
For now as long as we only have atom ("leaf") types and array type, if you have "an array of arrays" we can always flatten the two types into one. The problem starts to occur if you want to support record types, with records with 2+ fields where one is an array and another is something else, and then you want to make an array of that record type. Then you can't flatten any more and the type protos must become recursively defined. (although even though the type is recursively defined the particular type instances will merely be trees)

---

Reviewed 4 of 8 files at r2.
Review status: 8 of 16 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-:-KW_HP-y-rMcoTTYsGCx:b5y11s2)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
I'm also in favor of the "simple descriptor" approach as previously mentioned. @nvanbenschoten: any other thoughts here?

---

Review status: 8 of 16 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-:-KW_t2jjvAKNDye5ICbt:b-1j6yix)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: 8 of 16 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/pg_catalog.go, line 1138 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KWBUj1c83_iHU92vfzP:-KWdLp0XJX02D2rJn0F6:byzctu4) ([raw file](https://github.com/cockroachdb/cockroach/blob/712cc60a0f7506bca289d9a61f8efb5ee8c183bd/pkg/sql/pg_catalog.go#L1138)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> It seems to, though I'm not clear what the symptoms of this *not* working would be.
</details>
Won't `reflect.TypeOf(parser.TypeStringArray)` and `reflect.TypeOf(parser.TypeIntArray)` give you the same type? Same thing down below.

---

*[pkg/sql/parser/sql.y, line 400 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KWBVxN1NhEpqax2RXq8:-KWdMDh7-Yigh-iKRCGo:bobwssq) ([raw file](https://github.com/cockroachdb/cockroach/blob/712cc60a0f7506bca289d9a61f8efb5ee8c183bd/pkg/sql/parser/sql.y#L400)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> I don't understand your comment (though that's probably due to my rustiness with YACC). I'm looking at these:
> 
> https://github.com/postgres/postgres/blob/8c48375e5f43ebd832f93c9166d1fe0e639ff806/src/backend/parser/gram.y#L364
> https://github.com/postgres/postgres/blob/8c48375e5f43ebd832f93c9166d1fe0e639ff806/src/backend/parser/gram.y#L11158
> 
> and seeing that in the Postgres grammar, `opt_array_bounds` are simply lists of integers. Or, am I misinterpreting something here?
</details>
This should be `%type <Exprs> opt_array_bounds`, not `%type <bool> opt_array_bounds`. We're going to need that for the second branch of the rule.

---

*[pkg/sql/sqlbase/structured.proto, line 41 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KWBXBUZulFHdV1s0y-5:-KWdO8Itr5HxzimPRcjX:b7jejht) ([raw file](https://github.com/cockroachdb/cockroach/blob/712cc60a0f7506bca289d9a61f8efb5ee8c183bd/pkg/sql/sqlbase/structured.proto#L41)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Ok I want to react to this point specifically.
> There is no doubt that what we want is (like pg) homogeneous arrays, where all elements have the same type and "every row has the same number of columns" (the general property is to be multi-dimensional rectangularity, i.e. each dimension has the same size at every position in the other dimensions).
> 
> To achieve this, there are two implementation directions. 
> In both cases, there is a single array in memory that carries all the values.
> The difference is how you store the "metadata".
> 
> One direction is to have a single descriptor containing two pointers; one pointer to the full value array and one pointer to an array of sizes, one per dimension. 
> 
> The other direction is to have nested descriptors; each descriptor corresponds to a "type"; each descriptor contains 3 things, 1) a pointer to a slice, 2) a pointer to the type of the elements (which can be another array type), and a *stride*: an integer value used by the indexing function to multiply by a coordinate at that level to access one element in the slice.
> 
> The advantage of the first method is that it is slightly cheaper in memory usage when you have a large number of dimensions.
> The advantage of the second method is that the code for indexing is simpler.
> 
> Time-complexity wise they come to about the same. I think at this point it is up to the person who will implement this to decide how to do it.
</details>
In that case, it sounds like all we need is:
```
repeated int32 arrayDimensions = 5;
```
We can still use `kind` for the parameterized type of arrays, and `len(arrayDimensions) = 0` to indicate non-array types. Each dimension needs to be able to represent present and non-present size limits, so `uint32` alone won't work. If we make it an `int32`, we can use `-1` to represent no size limit.


Also, I'm a little concerned about making sure no-one is able to use this yet since we don't yet support persisting arrays. Could you add some tests that verify that it isn't actually possible to create a table with `[]int` columns?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584)*
<!-- Sent from Reviewable.io -->

Commit:
sql: fix pg_constraint to use int[] columns

Previously, the `conkey` and `confkey` columns of
`pg_catalog.pg_constraint` were creating mocked integer arrays. This
changes them to return real arrays.

Resolves #10523

cuongdo commented:
I've addressed this round of comments, fixed tests, rebased, and added new tests. PTAL.

---

Review status: 1 of 25 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/pg_catalog.go, line 1138 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KWBUj1c83_iHU92vfzP:-KWdnM3uwbRW-MM2CJZx:b1wi8ml) ([raw file](https://github.com/cockroachdb/cockroach/blob/712cc60a0f7506bca289d9a61f8efb5ee8c183bd/pkg/sql/pg_catalog.go#L1138)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Won't `reflect.TypeOf(parser.TypeStringArray)` and `reflect.TypeOf(parser.TypeIntArray)` give you the same type? Same thing down below.
</details>
Yes, though it seems like that doesn't do any harm in practice (and it certainly builds and runs). When `reflect.TypeOf(parser.TypeStringArray)` and `reflect.TypeOf(parser.TypeIntArray)` return the same type, the map contains a single value for that shared type.  If, at some later point, these two expressions return different types, the map will correctly contain two separate entries.

Here's some example code that demonstrates the map behavior:

https://play.golang.org/p/YZpC-dVqQg

---

*[pkg/sql/parser/sql.y, line 400 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KWBVxN1NhEpqax2RXq8:-KWdokgSKs5-IldZ9x-c:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/712cc60a0f7506bca289d9a61f8efb5ee8c183bd/pkg/sql/parser/sql.y#L400)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> This should be `%type <Exprs> opt_array_bounds`, not `%type <bool> opt_array_bounds`. We're going to need that for the second branch of the rule.
</details>
Done.

---

*[pkg/sql/sqlbase/structured.proto, line 41 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KWBXBUZulFHdV1s0y-5:-KWpR1dsQOh7WFoWxtKn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/712cc60a0f7506bca289d9a61f8efb5ee8c183bd/pkg/sql/sqlbase/structured.proto#L41)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> In that case, it sounds like all we need is:
> ```
> repeated int32 arrayDimensions = 5;
> ```
> We can still use `kind` for the parameterized type of arrays, and `len(arrayDimensions) = 0` to indicate non-array types. Each dimension needs to be able to represent present and non-present size limits, so `uint32` alone won't work. If we make it an `int32`, we can use `-1` to represent no size limit.
> 
> 
> Also, I'm a little concerned about making sure no-one is able to use this yet since we don't yet support persisting arrays. Could you add some tests that verify that it isn't actually possible to create a table with `[]int` columns?
</details>
Done.

---

*[pkg/sql/sqlbase/table.go, line 119 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KWpROkYESxW_bkcQ-7v:-KWpROkYESxW_bkcQ-7w:b-yqnx0r) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ebb5528e758c06c960e44fc697b7e1642d31dc1/pkg/sql/sqlbase/table.go#L119)):*
> ```Go
> 		col.Type.Kind = ColumnType_INT_ARRAY
> 		colDatumType = parser.TypeIntArray
> 		for i, e := range t.BoundsExprs {
> ```

I'm vaguely uncomfortable with the contents of this `for` loop. Is there a better way to get the final values of the `BoundsExprs`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-:-KWpRdpt5v9ZbkHlIZaR:bigyffc)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Reviewed 1 of 8 files at r1, 23 of 24 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[pkg/sql/pg_catalog.go, line 1138 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KWBUj1c83_iHU92vfzP:-KX4JmYN6yqGSR70Mfu0:b-5t187m) ([raw file](https://github.com/cockroachdb/cockroach/blob/712cc60a0f7506bca289d9a61f8efb5ee8c183bd/pkg/sql/pg_catalog.go#L1138)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> Yes, though it seems like that doesn't do any harm in practice (and it certainly builds and runs). When `reflect.TypeOf(parser.TypeStringArray)` and `reflect.TypeOf(parser.TypeIntArray)` return the same type, the map contains a single value for that shared type.  If, at some later point, these two expressions return different types, the map will correctly contain two separate entries.
> 
> Here's some example code that demonstrates the map behavior:
> 
> https://play.golang.org/p/YZpC-dVqQg
</details>
Sure, but that isn't the case for `datumToOid`, which is used to send back the correct OID.

---

*[pkg/sql/pg_catalog.go, line 424 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KX4L251yPjrbImNu7vH:-KX4L251yPjrbImNu7vI:b5kub03) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ebb5528e758c06c960e44fc697b7e1642d31dc1/pkg/sql/pg_catalog.go#L424)):*
> ```Go
> 						contype = conTypePKey
> 						conindid = h.IndexOid(db, table, c.Index)
> 						var err error
> ```

nit: I'd add a new line above each `err` declaration

---

*[pkg/sql/parser/builtins.go, line 1199 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KX4LQx-EZCG-R8aWLUN:-KX4LQx-EZCG-R8aWLUO:b6dv4qt) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ebb5528e758c06c960e44fc697b7e1642d31dc1/pkg/sql/parser/builtins.go#L1199)):*
> ```Go
> 				dimen := *args[1].(*DInt)
> 				// We do not currently support multi-dimensional arrays.
> 				if dimen != 1 || len(arr.Array) == 0 {
> ```

Why don't we add a `Len` method to `DArray`.

---

*[pkg/sql/parser/datum.go, line 1498 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KX4NVrqc3Z5OD3jIUm2:-KX4NVrqc3Z5OD3jIUm3:b-vyrqdi) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ebb5528e758c06c960e44fc697b7e1642d31dc1/pkg/sql/parser/datum.go#L1498)):*
> ```Go
> 		return TypeStringArray
> 	default:
> 		panic(fmt.Sprintf("invalid type for DArray: %s", d.ParamTyp))
> ```

`s/invalid/unsupported`

---

*[pkg/sql/parser/datum.go, line 1512 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KX4WHouOltCQ7E9rKix:-KX4WHov7NuHyZPBRizB:b-17ioza) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ebb5528e758c06c960e44fc697b7e1642d31dc1/pkg/sql/parser/datum.go#L1512)):*
> ```Go
> 		panic(makeUnsupportedComparisonMessage(d, other))
> 	}
> 	if d.ParamTyp != v.ParamTyp {
> ```

Do we need this? Won't the recursive call to `Compare` take care of it?

---

*[pkg/sql/parser/datum.go, line 1573 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KX4Nctl7YU2q74aFCKY:-KX4Nctl7YU2q74aFCKZ:boctamg) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ebb5528e758c06c960e44fc697b7e1642d31dc1/pkg/sql/parser/datum.go#L1573)):*
> ```Go
> 	n.Array = make([]Datum, len(d.Array))
> 	copy(n.Array, d.Array)
> 	for i := len(n.Array) - 1; i >= 0; i-- {
> ```

Could you open an issue to change this? The next value in an array should be the array appended with the minimum value of the parameterized data type (like strings or bytes), because they are not fixed length.

---

*[pkg/sql/parser/datum.go, line 1658 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KX4OIbDBGiwy2P3FeM5:-KX4OIbDBGiwy2P3FeM6:bvpe1m0) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ebb5528e758c06c960e44fc697b7e1642d31dc1/pkg/sql/parser/datum.go#L1658)):*
> ```Go
> func (d *DArray) Append(v Datum) error {
> 	switch v := v.(type) {
> 	case *DString:
> ```

This seems pretty crazy to me. We're taking a `Datum` and based on its type we are: dereferencing it, calling a specific method, and then allocating a whole new datum to append? Why not just perform the `ParamTyp` check with `.Equal` and if that works, append the datum directly? This should be a 5 line method and the other two shouldn't exist.

As a more general comment, a lot of these methods are now checking for `DString` or `DInt`. I understand that those are all we're supporting right now, but methods like `Append`, `ResolvedType`, and `Eval` don't need to know or worry about that. They can be implemented generically and correctly now without any problem. In fact, very little int the `sql/parser` package should be worrying about this, or we're just hurting ourselves by limiting future extensibility.

---

*[pkg/sql/parser/eval.go, line 2338 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KX4QZbXuDmFhgCzac1r:-KX4QZbYgCvooLCKoSfR:b-dnmfq) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ebb5528e758c06c960e44fc697b7e1642d31dc1/pkg/sql/parser/eval.go#L2338)):*
> ```Go
> func (t *Array) Eval(ctx *EvalContext) (Datum, error) {
> 	var array *DArray
> 	switch t.typ {
> ```

With respect to my comment above, use `colTypeToTypeAndValidArgTypes` here and dont worry about the datum type.

---

*[pkg/sql/parser/sql.y, line 2969 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KX4RvK2RIq_fJyKxibZ:-KX4RvK2RIq_fJyKxib_:b249gcf) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ebb5528e758c06c960e44fc697b7e1642d31dc1/pkg/sql/parser/sql.y#L2969)):*
> ```Yacc
>   simple_typename opt_array_bounds
>   {
>     if $2.val != nil {
> ```

`if exprs := $2.exprs(); exprs != nil {`

---

*[pkg/sql/parser/sql.y, line 2987 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KX4RqDl60V0cnyZhSAg:-KX4RqDl60V0cnyZhSAh:b-7ciqct) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ebb5528e758c06c960e44fc697b7e1642d31dc1/pkg/sql/parser/sql.y#L2987)):*
> ```Yacc
>   opt_array_bounds '[' ']' { $$.val = Exprs{NewDInt(DInt(-1))} }
> | opt_array_bounds '[' ICONST ']' { return unimplementedWithIssue(sqllex, 2115) }
> | /* EMPTY */ { $$.val = nil }
> ```

`Exprs(nil)`, see the comment above `sqlSymUnion`

---

*[pkg/sql/parser/type.go, line 315 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KX4SafobMKS_aWaF5p4:-KX4SafobMKS_aWaF5p5:b-1i8qi4) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ebb5528e758c06c960e44fc697b7e1642d31dc1/pkg/sql/parser/type.go#L315)):*
> ```Go
> // Equal implements the Type interface.
> func (a tArray) Equal(other Type) bool {
> 	oa, ok := other.(tArray)
> ```

You can simplify to:
```
if u, ok := other.(tArray); ok {
    return a.Typ.Equal(u.Typ)
}
return false
```

---

*[pkg/sql/parser/type_check.go, line 648 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KX4SmcOugWWAGrIgCcx:-KX4SmcOugWWAGrIgCcy:b-xnz1cl) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ebb5528e758c06c960e44fc697b7e1642d31dc1/pkg/sql/parser/type_check.go#L648)):*
> ```Go
> 	}
> 
> 	switch typ {
> ```

Yeah, this is one of the only places we need to be making the parameterized type restriction.

---

*[pkg/sql/parser/type_check_test.go, line 123 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KX4SyozcBn5jyhfYdFv:-KX4SyozcBn5jyhfYdFw:b-kx44cu) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc005bea58c1483f4b54d06550cce0898862fe78/pkg/sql/parser/type_check_test.go#L123)):*
> ```Go
> 		{`COALESCE(1, 2, 3, 4, '5')`, `incompatible COALESCE expressions: expected 1 to be of type string, found type int`},
> 		{`ARRAY[]`, `cannot determine type of empty array`},
> 		{`ARRAY[1, 2, 3]`, `unhandled parameterized array type parser.tInt`},
> ```

Let's leave this test case in for a type that we still don't support.

---

*[pkg/sql/sqlbase/structured.proto, line 41 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KWBXBUZulFHdV1s0y-5:-KX4TPmaUbXSxS7ARsyz:bdynjsr) ([raw file](https://github.com/cockroachdb/cockroach/blob/712cc60a0f7506bca289d9a61f8efb5ee8c183bd/pkg/sql/sqlbase/structured.proto#L41)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> Done.
</details>
But we still have `INT_ARRAY` as a kind, which is duplicating information. The idea was to use `array_dimensions` to signify if the `ColumnType` was an array of `kind` or just a single `kind`. Do we really want to add a new variant to the `Kind` enum for every type of array?

---

*[pkg/sql/sqlbase/structured.proto, line 49 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KX4Ti5XCjyFG2rPfnvF:-KX4Ti5XCjyFG2rPfnvG:bxlhkud) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ebb5528e758c06c960e44fc697b7e1642d31dc1/pkg/sql/sqlbase/structured.proto#L49)):*
> ```protobuf
>   // FLOAT and DECIMAL.
>   optional int32 precision = 3 [(gogoproto.nullable) = false];
>   repeated int32 arrayDimensions = 4;
> ```

`s/arrayDimensions/array_dimensions`. Also, this deserves some kind of comment.

---

*[pkg/sql/sqlbase/table.go, line 117 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KX4V8se1TyKSkEJKOGa:-KX4V8se1TyKSkEJKOGb:b-n8ofu9) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ebb5528e758c06c960e44fc697b7e1642d31dc1/pkg/sql/sqlbase/table.go#L117)):*
> ```Go
> 		colDatumType = parser.TypeBytes
> 	case *parser.ArrayColType:
> 		col.Type.Kind = ColumnType_INT_ARRAY
> ```

Shouldn't we reference `t.ColumnType` here?

---

*[pkg/sql/sqlbase/table.go, line 119 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KWpROkYESxW_bkcQ-7v:-KX4VFraF6AWp7UejMce:b-mejgiz) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ebb5528e758c06c960e44fc697b7e1642d31dc1/pkg/sql/sqlbase/table.go#L119)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> I'm vaguely uncomfortable with the contents of this `for` loop. Is there a better way to get the final values of the `BoundsExprs`?
</details>
This seems ok, but I'd use `parser.TypeCheckAndRequire`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
*[pkg/sql/sqlbase/structured.proto, line 41 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KWBXBUZulFHdV1s0y-5:-KX6p4nPguH5kG7Pcnjf:bth3sl3) ([raw file](https://github.com/cockroachdb/cockroach/blob/712cc60a0f7506bca289d9a61f8efb5ee8c183bd/pkg/sql/sqlbase/structured.proto#L41)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> But we still have `INT_ARRAY` as a kind, which is duplicating information. The idea was to use `array_dimensions` to signify if the `ColumnType` was an array of `kind` or just a single `kind`. Do we really want to add a new variant to the `Kind` enum for every type of array?
</details>
What you're suggesting is more elegant. However, `(k ColumnType_Kind) ToDatumType()` is supposed to return a `parser.Type` given just a `ColumnType_Kind`. That's not possible if the return value depends on both `ColumnType_Kind` and `ColumnType_ArrayDimensions`, but that'll require a fairly extensive refactoring of various methods that currently just take in a `ColumnType_Kind`. What do you suggest?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
Review status: 10 of 25 files reviewed at latest revision, 17 unresolved discussions, some commit checks pending.

---

*[pkg/sql/pg_catalog.go, line 1138 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KWBUj1c83_iHU92vfzP:-KX6TVP0NiACARcLlLRu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/712cc60a0f7506bca289d9a61f8efb5ee8c183bd/pkg/sql/pg_catalog.go#L1138)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Sure, but that isn't the case for `datumToOid`, which is used to send back the correct OID.
</details>
Done.

---

*[pkg/sql/pg_catalog.go, line 424 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KX4L251yPjrbImNu7vH:-KX6RRXeiup1bvKD4v8T:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ebb5528e758c06c960e44fc697b7e1642d31dc1/pkg/sql/pg_catalog.go#L424)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> nit: I'd add a new line above each `err` declaration
</details>
Done.

---

*[pkg/sql/parser/builtins.go, line 1199 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KX4LQx-EZCG-R8aWLUN:-KX6UqadcRK-LJDTmEVE:b831u67) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ebb5528e758c06c960e44fc697b7e1642d31dc1/pkg/sql/parser/builtins.go#L1199)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Why don't we add a `Len` method to `DArray`.
</details>
It actually already exists; I just forgot about it. Switched instances `len(arr.Array)` over to `arr.Len()`.

---

*[pkg/sql/parser/datum.go, line 1498 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KX4NVrqc3Z5OD3jIUm2:-KX6V934Ead1EMn4w4t-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ebb5528e758c06c960e44fc697b7e1642d31dc1/pkg/sql/parser/datum.go#L1498)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> `s/invalid/unsupported`
</details>
Done.

---

*[pkg/sql/parser/datum.go, line 1512 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KX4WHouOltCQ7E9rKix:-KX6VNvfOROeDjmGxped:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ebb5528e758c06c960e44fc697b7e1642d31dc1/pkg/sql/parser/datum.go#L1512)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Do we need this? Won't the recursive call to `Compare` take care of it?
</details>
Done.

---

*[pkg/sql/parser/datum.go, line 1573 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KX4Nctl7YU2q74aFCKY:-KX6W_c6jGLJOroSs5xo:b-eb2nty) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ebb5528e758c06c960e44fc697b7e1642d31dc1/pkg/sql/parser/datum.go#L1573)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Could you open an issue to change this? The next value in an array should be the array appended with the minimum value of the parameterized data type (like strings or bytes), because they are not fixed length.
</details>
Done: #10902 

---

*[pkg/sql/parser/datum.go, line 1658 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KX4OIbDBGiwy2P3FeM5:-KX6_ejoSKLzrvUQyfLt:b-1xjyuj) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ebb5528e758c06c960e44fc697b7e1642d31dc1/pkg/sql/parser/datum.go#L1658)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> This seems pretty crazy to me. We're taking a `Datum` and based on its type we are: dereferencing it, calling a specific method, and then allocating a whole new datum to append? Why not just perform the `ParamTyp` check with `.Equal` and if that works, append the datum directly? This should be a 5 line method and the other two shouldn't exist.
> 
> As a more general comment, a lot of these methods are now checking for `DString` or `DInt`. I understand that those are all we're supporting right now, but methods like `Append`, `ResolvedType`, and `Eval` don't need to know or worry about that. They can be implemented generically and correctly now without any problem. In fact, very little int the `sql/parser` package should be worrying about this, or we're just hurting ourselves by limiting future extensibility.
</details>
Since I don't know the surrounding code thoroughly, I was focusing on correctness over avoiding allocations. Since `Datum` are generally pointers to structs underneath, I was hesitant to pass mutable state around. I was also wary of unintentional sharing of mutable `Datum`. I've fixed this.

I've also removed most of the type-aware code in `sql/parser`.

---

*[pkg/sql/parser/eval.go, line 2338 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KX4QZbXuDmFhgCzac1r:-KX6jrnWZYe3EUN_UwmC:buu6xtx) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ebb5528e758c06c960e44fc697b7e1642d31dc1/pkg/sql/parser/eval.go#L2338)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> With respect to my comment above, use `colTypeToTypeAndValidArgTypes` here and dont worry about the datum type.
</details>
I used a different approach, since I couldn't see how to make `colTypeToTypeAndValidArgTypes` to construct a properly parameterized `DArray`.

---

*[pkg/sql/parser/sql.y, line 2969 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KX4RvK2RIq_fJyKxibZ:-KX6kNIjUal49oCwDH6B:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ebb5528e758c06c960e44fc697b7e1642d31dc1/pkg/sql/parser/sql.y#L2969)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> `if exprs := $2.exprs(); exprs != nil {`
</details>
Done.

---

*[pkg/sql/parser/sql.y, line 2987 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KX4RqDl60V0cnyZhSAg:-KX6kb3n0NfUUhcQEL7_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ebb5528e758c06c960e44fc697b7e1642d31dc1/pkg/sql/parser/sql.y#L2987)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> `Exprs(nil)`, see the comment above `sqlSymUnion`
</details>
Done.

---

*[pkg/sql/parser/type.go, line 315 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KX4SafobMKS_aWaF5p4:-KX6x9RJRtLQVNzGL2in:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ebb5528e758c06c960e44fc697b7e1642d31dc1/pkg/sql/parser/type.go#L315)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> You can simplify to:
> ```
> if u, ok := other.(tArray); ok {
>     return a.Typ.Equal(u.Typ)
> }
> return false
> ```
</details>
Done.

---

*[pkg/sql/parser/type_check_test.go, line 123 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KX4SyozcBn5jyhfYdFv:-KX6mhfcoyPJh_8MFca0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc005bea58c1483f4b54d06550cce0898862fe78/pkg/sql/parser/type_check_test.go#L123)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Let's leave this test case in for a type that we still don't support.
</details>
Done.

---

*[pkg/sql/sqlbase/structured.proto, line 49 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KX4Ti5XCjyFG2rPfnvF:-KX6pqkBpG9s-jZtcq_W:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ebb5528e758c06c960e44fc697b7e1642d31dc1/pkg/sql/sqlbase/structured.proto#L49)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> `s/arrayDimensions/array_dimensions`. Also, this deserves some kind of comment.
</details>
Done.

---

*[pkg/sql/sqlbase/table.go, line 117 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KX4V8se1TyKSkEJKOGa:-KX6sPdXd2Fmw6SFDs3P:b-30t9k5) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ebb5528e758c06c960e44fc697b7e1642d31dc1/pkg/sql/sqlbase/table.go#L117)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Shouldn't we reference `t.ColumnType` here?
</details>
`t.columnType` is unexported, so I've gone with an alternative.

---

*[pkg/sql/sqlbase/table.go, line 119 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KWpROkYESxW_bkcQ-7v:-KX6wofuv-Bao7LAFb7H:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ebb5528e758c06c960e44fc697b7e1642d31dc1/pkg/sql/sqlbase/table.go#L119)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> This seems ok, but I'd use `parser.TypeCheckAndRequire`.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Reviewed 3 of 15 files at r6, 1 of 1 files at r7.
Review status: 14 of 25 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/sql/pgtypes.go, line 76 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KX8GK8dwANawwKWzZpB:-KX8GK8e99GBl27ScmuW:b-j3q9ei) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cfa075b77444536ab9985426a09ce6a2ca32b4e/pkg/sql/pgtypes.go#L76)):*
> ```Go
> 	dOid, ok := datumToOid[reflect.TypeOf(typ)]
> 	if !ok {
> 		if typ == parser.TypeIntArray {
> ```

This should be a switch statement. Also, could you leave a TODO to remove `datumToOid` and just use a switch statement for everything instead?

---

*[pkg/sql/parser/datum.go, line 1573 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KX4Nctl7YU2q74aFCKY:-KX8GsXm7wbTe5GK0_Cj:b-zcjvte) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ebb5528e758c06c960e44fc697b7e1642d31dc1/pkg/sql/parser/datum.go#L1573)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> Done: #10902 
</details>
Thanks

---

*[pkg/sql/parser/datum.go, line 1658 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KX4OIbDBGiwy2P3FeM5:-KX8GwR_RKIrRXJtjQnf:b-qgcho8) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ebb5528e758c06c960e44fc697b7e1642d31dc1/pkg/sql/parser/datum.go#L1658)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> Since I don't know the surrounding code thoroughly, I was focusing on correctness over avoiding allocations. Since `Datum` are generally pointers to structs underneath, I was hesitant to pass mutable state around. I was also wary of unintentional sharing of mutable `Datum`. I've fixed this.
> 
> I've also removed most of the type-aware code in `sql/parser`.
</details>
Yep, this looks a lot better!

---

*[pkg/sql/parser/datum.go, line 1505 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KX8GoY9JRPM91cBzjav:-KX8GoY9JRPM91cBzjaw:ba7314k) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cfa075b77444536ab9985426a09ce6a2ca32b4e/pkg/sql/parser/datum.go#L1505)):*
> ```Go
> 		panic(makeUnsupportedComparisonMessage(d, other))
> 	}
> 	n := len(d.Array)
> ```

`d.Len()` throughout.

---

*[pkg/sql/parser/sql.y, line 3028 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KX8PbM5j1UQM0KWcdPZ:-KX8PbM6PyAHPraM-Dkp:b-kw0d1p) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cfa075b77444536ab9985426a09ce6a2ca32b4e/pkg/sql/parser/sql.y#L3028)):*
> ```Yacc
>     if exprs := $2.exprs(); exprs != nil {
>       var err error
>       $$.val, err = arrayOf($1.colType(), $2.exprs())
> ```

You can use `exprs` here now

---

*[pkg/sql/sqlbase/structured.proto, line 41 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KWBXBUZulFHdV1s0y-5:-KX8PvEYitEXtnwPNlV5:bapow6r) ([raw file](https://github.com/cockroachdb/cockroach/blob/712cc60a0f7506bca289d9a61f8efb5ee8c183bd/pkg/sql/sqlbase/structured.proto#L41)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> What you're suggesting is more elegant. However, `(k ColumnType_Kind) ToDatumType()` is supposed to return a `parser.Type` given just a `ColumnType_Kind`. That's not possible if the return value depends on both `ColumnType_Kind` and `ColumnType_ArrayDimensions`, but that'll require a fairly extensive refactoring of various methods that currently just take in a `ColumnType_Kind`. What do you suggest?
</details>
I guess this is fine for now then, but could we bump this up to `= 100` and leave a big TODO about not allowing the persistence of these Kinds until we fully resolve this question. I'm afraid of painting ourselves into a corner if we're not careful here and suddenly need to preserve backwards compatibility with decisions we're making now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
Review status: 10 of 25 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/sql/pgtypes.go, line 76 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KX8GK8dwANawwKWzZpB:-KX8n1gzLllLCJRQdWN1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cfa075b77444536ab9985426a09ce6a2ca32b4e/pkg/sql/pgtypes.go#L76)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> This should be a switch statement. Also, could you leave a TODO to remove `datumToOid` and just use a switch statement for everything instead?
</details>
Done.

---

*[pkg/sql/parser/datum.go, line 1505 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KX8GoY9JRPM91cBzjav:-KX8nMXnyI7RBQ_CcS_4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cfa075b77444536ab9985426a09ce6a2ca32b4e/pkg/sql/parser/datum.go#L1505)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> `d.Len()` throughout.
</details>
Done.

---

*[pkg/sql/parser/sql.y, line 3028 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KX8PbM5j1UQM0KWcdPZ:-KX8pmNxvnizfhCkpmCB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cfa075b77444536ab9985426a09ce6a2ca32b4e/pkg/sql/parser/sql.y#L3028)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> You can use `exprs` here now
</details>
Done.

---

*[pkg/sql/sqlbase/structured.proto, line 41 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-KWBXBUZulFHdV1s0y-5:-KX8rU8zhjvrw8fEbRgl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/712cc60a0f7506bca289d9a61f8efb5ee8c183bd/pkg/sql/sqlbase/structured.proto#L41)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> I guess this is fine for now then, but could we bump this up to `= 100` and leave a big TODO about not allowing the persistence of these Kinds until we fully resolve this question. I'm afraid of painting ourselves into a corner if we're not careful here and suddenly need to preserve backwards compatibility with decisions we're making now.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 10 of 25 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10584#-:-KXAW4mh7QPxJyJMFkuW:bnfp4nl)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
Thanks for your thorough, patient reviews!
